src/Nethermind/Nethermind.Taiko/Rpc/SurgeGasPriceOracle.cs

```diff
@@ -0,0 +1,256 @@
+// SPDX-FileCopyrightText: 2025 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+using Nethermind.Blockchain.Find;
+using Nethermind.Core;
+using Nethermind.Core.Crypto;
+using Nethermind.Core.Specs;
+using Nethermind.Int256;
+using Nethermind.JsonRpc.Client;
+using Nethermind.JsonRpc.Modules.Eth.GasPrice;
+using Nethermind.Logging;
+using Nethermind.Taiko.Config;
+
+namespace Nethermind.Taiko.Rpc;
+
+public class SurgeGasPriceOracle : GasPriceOracle
+{
+    private static readonly string _className = nameof(SurgeGasPriceOracle);
+    private readonly IJsonRpcClient _l1RpcClient;
+    private readonly ISurgeConfig _surgeConfig;
+    private readonly GasUsageRingBuffer _gasUsageBuffer;
+
+    private const int BlobSize = 128 * 1024;
+
+    public SurgeGasPriceOracle(
+        IBlockFinder blockFinder,
+        ILogManager logManager,
+        ISpecProvider specProvider,
+        UInt256 minGasPrice,
+        IJsonRpcClient l1RpcClient,
+        ISurgeConfig surgeConfig) : base(blockFinder, specProvider, logManager, minGasPrice)
+    {
+        _l1RpcClient = l1RpcClient;
+        _surgeConfig = surgeConfig;
+        _gasUsageBuffer = new GasUsageRingBuffer(_surgeConfig.L2GasUsageWindowSize);
+    }
+
+    private UInt256 FallbackGasPrice() => _gasPriceEstimation.LastPrice ?? _minGasPrice;
+
+    public override UInt256 GetGasPriceEstimate()
+    {
+        Block? headBlock = _blockFinder.Head;
+        if (headBlock is null)
+        {
+            if (_logger.IsTrace) _logger.Trace($"[{_className}] No head block available, using fallback gas price");
+            return FallbackGasPrice();
+        }
+
+        Hash256 headBlockHash = headBlock.Hash!;
+        if (_gasPriceEstimation.TryGetPrice(headBlockHash, out UInt256? price))
+        {
+            if (_logger.IsTrace) _logger.Trace($"[{_className}] Using cached gas price estimate: {price}");
+            return price!.Value;
+        }
+
+        // Get the fee history from the L1 client with RPC
+        L1FeeHistoryResults? feeHistory = GetL1FeeHistory().GetAwaiter().GetResult();
```

Comment:
Here and in other places, it might be better to await instead of .GetAwaiter().GetResult();